### PR TITLE
add l1 sparse attack

### PR DESCRIPTION
**Pull request by LaurentMnr95**
add l1 sparse attack from https://arxiv.org/pdf/1904.13000.pdf

### CODE REVIEW

**Comment by gwding** [comment]
@LaurentMnr95 Thanks for contributing! I'm travelling this week, and I'll review this PR whenever I got a chance. but it might take a few days.

File: advertorch/utils.py [hunk 1]
```diff
@@ -156,6 +156,48 @@ def _get_norm_batch(x, p):
...
+    sign = x.sign()
+
+    mag = abs_input - lamda
+    mag = (torch.abs(mag) + mag) / 2
+
+    return mag * sign
+
+
+def batch_l1_proj_flat(x, eps=1):
+    # this function does not return the good result when x is a cuda tensor
```
Comment (gwding) [inline advertorch/utils.py]: Can you elaborate a bit on this?

File: advertorch/utils.py [hunk 2]
```diff
@@ -156,6 +156,48 @@ def _get_norm_batch(x, p):
...
+    abs_input = torch.abs(x)
+    sign = x.sign()
+
+    mag = abs_input - lamda
+    mag = (torch.abs(mag) + mag) / 2
+
+    return mag * sign
+
+
+def batch_l1_proj_flat(x, eps=1):
```
Comment (gwding) [inline advertorch/utils.py]: I'm trying to understand the projection algorithm here. in https://arxiv.org/pdf/1904.13000.pdf, it only mentions that the algorithm is from Duchi [8], which is this paper https://stanford.edu/~jduchi/projects/DuchiShSiCh08.pdf . In this paper, which algorithm is this function implementing?

File: advertorch/attacks/iterative_projected_gradient.py [hunk 3]
```diff
@@ -29,7 +30,7 @@
 
 def perturb_iterative(xvar, yvar, predict, nb_iter, eps, eps_iter, loss_fn,
                       delta_init=None, minimize=False, ord=np.inf,
-                      clip_min=0.0, clip_max=1.0):
+                      clip_min=0.0, clip_max=1.0, sparsity=0.95):
```
Comment (gwding) [inline advertorch/attacks/iterative_projected_gradient.py]: Let's name `sparsity` more specifically, e.g. `l1sparsity` or `sparsity_l1`, so people know it is only for L1

File: advertorch/attacks/iterative_projected_gradient.py [hunk 4]
```diff
@@ -77,10 +78,31 @@ def perturb_iterative(xvar, yvar, predict, nb_iter, eps, eps_iter, loss_fn,
...
+
+            batch_size = grad.size(0)
+            view = abs_grad.view(batch_size, -1)
+            view_size = view.size(1)
+            vals, idx = view.topk(int((1-sparsity)*view_size))
+
+            out = torch.zeros_like(view).scatter_(1, idx, vals)
+
+            out = out.view_as(grad)
+            grad = grad.sign()*(out > 0).float()
```
Comment (gwding) [inline advertorch/attacks/iterative_projected_gradient.py]: There's some python formatting issues here. I'll add a format test in the master branch (today or tomorrow), so these can be caught automatically.

File: advertorch/attacks/iterative_projected_gradient.py [hunk 5]
```diff
@@ -210,6 +233,31 @@ def __init__(
             clip_min, clip_max, ord, targeted)
 
 
+class SparseL1Attack(PGDAttack):
```
Comment (gwding) [inline advertorch/attacks/iterative_projected_gradient.py]: On top of the `SparseL1Attack`, also add an `L1PGDAttack` as a special case of it? so later people can easily compare them.

**Comment by gwding** [comment]
@LaurentMnr95 Thanks for contributing again! I left some comments at different locations, and there're a few things need to be resolved.

Besides, could you also add an example that reproduces some of the key results in the paper? as a performance check of the implementation? both on SparseL1DescentAttack and L1PGDAttack

File: advertorch/attacks/__init__.py [hunk 6]
```diff
@@ -21,6 +21,7 @@
 from .iterative_projected_gradient import PGDAttack
 from .iterative_projected_gradient import LinfPGDAttack
 from .iterative_projected_gradient import L2PGDAttack
+from .iterative_projected_gradient import SparseL1Attack
```
Comment (gwding) [inline advertorch/attacks/__init__.py]: maybe name as `SparseL1DescentAttack` following the paper?

File: advertorch/utils.py (hunk 2 above)
Comment (LaurentMnr95) [reply advertorch/utils.py]: I took inspiration from the following implementation from Duchi's paper https://gist.github.com/daien/1272551/edd95a6154106f8e28209a1c7964623ef8397246 and made it consistent with batch attacks.

**Comment by LaurentMnr95** [comment]
Many thanks for your review @gwding I am on a conference this week, I will try to answer your review
asap.

File: advertorch/utils.py (hunk 1 above)
Comment (LaurentMnr95) [reply advertorch/utils.py]: I corrected this in the next commit, the problem came from the "max" function that do not operate the same way on CUDA devices.

File: advertorch/attacks/iterative_projected_gradient.py (hunk 3 above)
Comment (LaurentMnr95) [reply advertorch/attacks/iterative_projected_gradient.py]: Done!

File: advertorch/attacks/__init__.py (hunk 6 above)
Comment (LaurentMnr95) [reply advertorch/attacks/__init__.py]: Done!

File: advertorch/attacks/iterative_projected_gradient.py (hunk 5 above)
Comment (LaurentMnr95) [reply advertorch/attacks/iterative_projected_gradient.py]: I added L1PGD attack!

File: advertorch/attacks/iterative_projected_gradient.py (hunk 4 above)
Comment (LaurentMnr95) [reply advertorch/attacks/iterative_projected_gradient.py]: Ok let me know when it has been made

Comment (LaurentMnr95) [reply advertorch/attacks/iterative_projected_gradient.py]: Normally it is ok with flake8 test

**Comment by LaurentMnr95** [comment]
I think I have corrected all the points you mentioned. If you have further concerns, don't hesitate!

**Comment by gwding** [comment]
@LaurentMnr95 Thanks for addressing the comments very quickly! I was trying to read the L1 projection paper and understand the algorithm, but seems it might take me sometime. will work on it and review further.

**Comment by LaurentMnr95** [comment]
Thank you @gwding, I tested the implementation on simple 2D cases and it returned the good results

File: advertorch/attacks/iterative_projected_gradient.py (hunk 4 above)
Comment (gwding) [reply advertorch/attacks/iterative_projected_gradient.py]: I'll merge the flake8 test branch after this current branch got merged.
Somehow I think `(1-l1_sparsity)*view_size` will be caught and needs to be corrected to `(1 - l1_sparsity) * view_size` with spaces, but it seems that flake8 does not catch that always, so I'll need some further investigation.

File: advertorch/utils.py [hunk 7]
```diff
@@ -156,6 +156,61 @@ def _get_norm_batch(x, p):
     return x.abs().pow(p).view(batch_size, -1).sum(dim=1).pow(1. / p)
 
 
+def soft_thresh(lamda, x):
```
Comment (gwding) [inline advertorch/utils.py]: I wasn't able to find why we need `soft_thresh`, could you explain a bit or point to the corresponding location in the paper?

File: advertorch/utils.py (hunk 2 above)
Comment (gwding) [reply advertorch/utils.py]: I have a better understanding of the algorithm by reading the paper. 
But it is still a little bit hard for me to follow the code. Would you mind to change the name of the variables to those used in `Figure 1. Algorithm for projection onto the simple` in https://stanford.edu/~jduchi/projects/DuchiShSiCh08.pdf , so it is easier for people to check and understand?

**Comment by gwding** [comment]
@LaurentMnr95 Had a another pass through the code after understanding the L1 projection method by reading the paper. 
Once these comments are resolved, I might do some style-related changes and make PR to your current branch, and if you're ok with those. We should be good to merge.

File: advertorch/utils.py (hunk 7 above)
Comment (LaurentMnr95) [reply advertorch/utils.py]: The soft threshold is needed on the last step of the projection: \max{w_i-\theta,0}  on the last step of the projection on the simplex.

File: advertorch/utils.py (hunk 2 above)
Comment (LaurentMnr95) [reply advertorch/utils.py]: I changed the name of variables to make it consistent with notations from Duchi's  paper

File: advertorch/utils.py [hunk 8]
```diff
@@ -156,6 +156,69 @@ def _get_norm_batch(x, p):
...
+        return x
+
+    # make the projection on l1 ball for elements outside the ball
+    view = x_b
+    view_size = view.size(1)
+    mu = view.abs().sort(1, descending=True)[0]
+    vv = torch.arange(view_size).float().to(x.device)
+    st = (mu.cumsum(1)-z)/(vv+1)
+    u = (mu-st) > 0
+    rho = (1-u).cumsum(dim=1).eq(0).sum(1)-1
```
Comment (gwding) [inline advertorch/utils.py]: Maybe my last question regarding the projection algorithm, in the paper, rho is calculated via max. So I'm a bit confused about how this online can be used to achieve the max over something.

**Comment by gwding** [comment]
I've made some small changes in https://github.com/LaurentMnr95/advertorch/pull/1 , if you think these are ok, you can merge them into your branch and this PR should be updated accordingly.

Besides those, my (hopefully) last question is here https://github.com/BorealisAI/advertorch/pull/24/files#r302207536

File: advertorch/utils.py (hunk 8 above)
Comment (LaurentMnr95) [reply advertorch/utils.py]: Actually, we need to get the first element that verify some conditions. On  CPUs, max are ordered and returned always the same way. On GPUs, the order is random, so this trick is made to get the first element  verifying the condition.

Comment (gwding) [reply advertorch/utils.py]: Thanks. I think I get it now. Maybe it's just me, the logic does seem to be a bit involved. We might want to refactor to simplify these calculations a bit, which might need your help in the future :)
For now, I think it is good to merge.

**Pull request merged**